src/dstack/_internal/server/services/pools.py

```diff
@@ -494,6 +511,53 @@ def filter_pool_instances(
     return instances
 
 
+def get_shared_pool_instances_with_offers(
+    pool_instances: List[InstanceModel],
+    profile: Profile,
+    requirements: Requirements,
+    *,
+    idle_only: bool = False,
+    fleet_model: Optional[FleetModel] = None,
+    volumes: Optional[List[List[Volume]]] = None,
+) -> list[tuple[InstanceModel, InstanceSharedOffer]]:
+    instances_with_offers: list[tuple[InstanceModel, InstanceSharedOffer]] = []
+    query_filter = requirements_to_query_filter(requirements)
+    filtered_instances = filter_pool_instances(
+        pool_instances=pool_instances,
+        profile=profile,
+        fleet_model=fleet_model,
+        multinode=False,
+        volumes=volumes,
+        shared=True,
+    )
+    for instance in filtered_instances:
+        if idle_only and instance.status not in [InstanceStatus.IDLE, InstanceStatus.BUSY]:
+            continue
+        offer = get_instance_offer(instance)
+        if offer is None:
+            continue
+        shared_info = get_instance_shared_info(instance)
+        if shared_info is None:
+            continue
+        total_blocks = shared_info.total_blocks
+        if total_blocks == "auto":
+            # provisioning in progress
+            continue
+        idle_blocks = total_blocks - shared_info.busy_blocks
+        for blocks in range(1, total_blocks + 1):
+            shared_offer = generate_shared_offer(offer, blocks, total_blocks)
+            catalog_item = offer_to_catalog_item(shared_offer)
+            if gpuhunt.matches(catalog_item, query_filter):
+                if blocks <= idle_blocks:
+                    shared_offer.availability = InstanceAvailability.IDLE
+                else:
+                    shared_offer.availability = InstanceAvailability.BUSY
+                if shared_offer.availability == InstanceAvailability.IDLE or not idle_only:
+                    instances_with_offers.append((instance, shared_offer))
+                break
```

Comment:
So we show only the smallest matching offer? Is it intended. Asking because the doc showed all possible offers.

Reply:
Yes, IMO there is no reason to show all virual offers as they won't be picked anyway